Use formatted output in crm_mon

clumens opened the pull request:
Here it is - the mega PR that adds formatted output support to crm_mon.  This is built on top of my other recent PR which makes it look a little larger than it would be, but it's still an awful lot of patches.  It is definitely possible I've missed something.  I'd like to encourage people to try this out and let me know where I've screwed up the formatting.  We'll likely be talking about this one for a while.

clumens commented:
A couple comments:

* I really dislike f5d3005 and to a lesser extent 4c21b17.  I don't know what to do about them, though.

* If it would make it easier to wade through this mega PR, I could create a smaller one that just adds all the formatted output messages but does not start using them.  I could then use this PR to make the changes to crm_mon.  So the first one would be almost entirely additive, and the second would be almost entirely removing.  That might make it easy to track what I imagine will be an awful lot of comments.

clumens commented on tools/crm_mon_output.c:
```diff
@@ -0,0 +1,436 @@
+        return curses_indented_printf;
+    } else {
+        return pcmk__indented_printf;
+    }
```
Fixing that up was way easier than I thought it'd be.  I must have thought it would be a lot more useful in the beginning, but it didn't end up that way.  The duplication is not too bad.

clumens commented on tools/crm_mon_print.c:
```diff
@@ -1249,18 +1246,10 @@ print_node_attributes(mon_state_t *state, pe_working_set_t *data_set, unsigned i
+    if (state->output_format == mon_output_xml) {
+        state->out->message(state->out, "begin-parent", "summary");
+    } else {
+        state->out->message(state->out, "header", 2, "Cluster Summary");
```
I originally thought the header message would be more broadly useful, but it turns out I'm only using it in one or two places.  So that can just go away entirely.  I'll look at further changing the cluster summary messages.

clumens commented:
I've split the last several commits out into their own PR.  When that one gets merged, I'll rebase this one and get rid of those commits.

clumens commented:
Updated to rebase on top of changes to master.  I think I got the recent Y2038 changes merged in correctly, but it's definitely worth a double check.

clumens commented:
I think I am now actually done pushing new commits and am ready to start working on any additional comments.

kgaillot commented on lib/pengine/clone.c:
```diff
@@ -993,11 +983,11 @@ pe__clone_text(pcmk__output_t *out, va_list args)
     }
 
-    free(child_text);
+    out->end_list(out);
```
I suspect at this point there's no point in having separate pe__short_output_*() and pe__clone_*() for text and html. We can probably get rid of pe__short_output_*() altogether.

clumens replied:
The pe__short_output_ functions are called by the pe__clone_ functions.  Should I inline what they were doing, or do we just not want that output anymore?

kgaillot replied:
I was thinking inlining

kgaillot commented on lib/pengine/native.c:
```diff
@@ -824,35 +818,38 @@ pe__common_output_text(pcmk__output_t *out, resource_t * rsc, const char *pre_te
+        total = g_strjoinv(" ", arr);
+        unames = crm_strdup_printf(" [ %s ]", total);
+
+        free(total);
```
And here's an example of why we try to avoid mixing and matching memory allocation :)

clumens replied:
Switched to using g_strfreev for the moment.  If we end up changing how the list is made, I'll change how it's freed too.

kgaillot replied:
g_strjoinv() needs g_free()

kgaillot commented on lib/pengine/pe_output.c:
```diff
@@ -101,28 +101,23 @@ pe__group_text(pcmk__output_t *out, va_list args)
+    char *buf = crm_strdup_printf("Resource Group: %s", rsc->id);
 
-    fprintf(out->dest, "%sResource Group: %s", pre_text, rsc->id);
+    out->begin_list(out, buf, NULL, NULL);
```
I wonder if it would be helpful (separate from this PR) if it were called as begin_list(out, NULL, NULL, "Resource Group: %s", rsc->id)

clumens replied:
Yeah that's a pretty good idea.  There's probably a couple more functions that could benefit from taking a list of varargs.  list_item immediately comes to mind.

clumens replied:
See: #1896.

kgaillot commented on tools/crm_mon_output.c:
```diff
@@ -827,6 +913,102 @@ node_attribute_xml(pcmk__output_t *out, va_list args) {
+
+    free(buf);
+    return 0;
+}
```
Can't this function be used for both text and html?

clumens replied:
Yep.

kgaillot commented on tools/crm_mon_output.c:
```diff
@@ -1019,6 +1042,74 @@ resource_brief_console(pcmk__output_t *out, va_list args) {
+    out->begin_list(out, buf, NULL, NULL);
+    free(buf);
+    return 0;
+}
```
Ditto

clumens replied:
Also yes.

clumens commented on tools/crm_mon_print.c:
```diff
@@ -507,7 +424,23 @@ print_rsc_history_start(mon_state_t *state, pe_working_set_t *data_set,
+                case mon_output_console:
+                case mon_output_plain:
+                    print_as(state->output_format, " %s",
+                             pcmk_format_named_time(CRM_LAST_FAILURE_PREFIX, last_failure));
```
I am positive I already did this.  I wonder if it somehow got lost in a rebase.  Oh well, guess I'll go do it again.

clumens replied:
I corrected several more of these.  For pcmk_format_named_time, there's nowhere it's being called where it is not freed, so this must have gotten fixed in an intermediate commit.  I am having some trouble tracking down which one it was.

clumens replied:
Okay, I went through and made sure each patch where one of these functions is called also has a free, so they should all be right now.  If something isn't getting freed in the patch where it's introduced, point it out and I'll take care of it.  Note there's still a patch that adds frees of time_t_string calls but that's because those calls don't come from this PR.

clumens commented on tools/crm_mon_print.c:
```diff
@@ -1243,42 +1238,10 @@ print_node_attributes(mon_state_t *state, pe_working_set_t *data_set, unsigned i
+        state->out->begin_list(state->out, "summary", NULL, NULL);
+    } else {
+        state->out->begin_list(state->out, "Cluster Summary", NULL, NULL);
     }
```
Not with the simple-list option set, which should be forced on if the output format is XML.

kgaillot replied:
Ah, I overlooked that. Just thinking out loud, since this is all just for backward compatibility and hopefully won't expand beyond this, what do you think of moving this to begin_list_xml() with e.g. if (simple_list) { if !strcmp(name, "Cluster Summary") name = "summary" else if ... }?

clumens replied:
Sorry, I'm not quite following you here.

kgaillot replied:
I hate when a caller needs to know the output format, since it goes against the design. In this case we need it for backward compatibility, but we shouldn't need such hacks outside of crm_mon going forward. So I'm wondering which is better -- putting the if/else in crm_mon, or moving it to xml_begin_list() like:

crm_mon:

    state->out->begin_list(state->out, "Cluster Summary", NULL, NULL);

output_xml.c:

    if (legacy_xml) {
        if (!strcmp(name, "Cluster Summary")) {
            name = "summary";
        else if ...
    }

Since crm_mon will be the only thing using legacy_xml, it shouldn't be a performance hit for any other callers. Of course it's ugly either way, it's just a question of whether the ugliness should be in crm_mon or output_xml.

kgaillot replied:
The argument for keeping it in crm_mon is so it doesn't clutter the general-purpose code, the argument for putting it in output_xml is that it's legacy-XML-specific handling.

clumens replied:
Oh, I see now.  There's a potential third way that accomplishes the same thing without cluttering up the general purpose code.  I could add a tools/crm_mon_xml.c (which we've successfully avoided so far) that redefines begin_list to do the name translation.

clumens replied:
Yeah, that seems to work pretty well.  It's basically just a new file with a big if-else block for all the possible names.  Everything is contained in that one spot.  One downside is there's no way to see if legacy_xml or simple_list are set (they're static to lib/common/output_xml.c) so crm_mon will be forced to just assume that legacy_xml is always true.  We force that in crm_mon.c so it's not really a big deal.

I think this works best as a new patch since it touches so many other patches.  Also, there's still several checks against output_format I cannot get rid of, though most are in crm_mon.c itself.

clumens commented on tools/crm_mon.c:
```diff
@@ -1085,7 +1079,7 @@ main(int argc, char **argv)
             } else {
                 if (output_format == mon_output_html) {
-                    print_as(output_format, "Writing html to %s ...\n", output_filename);
+                    printf("Writing html to %s ...\n", args->output_dest);
```
Hm, yes probably.  I don't think output_dest can ever be stdout for HTML at the moment but it's probably best to be ready.

kgaillot replied:
Wouldn't --output-as=html --output-to=- do it?

clumens replied:
It should, but it currently does not:

```
clumens@localhost:~/src/pacemaker$ CIB_file=cts/scheduler/anon-instance-pending.xml tools/crm_mon --output-as html --output-to -
lt-crm_mon: Missing argument for --output-to
Usage:
  lt-crm_mon [OPTION]
```

So I guess I need to figure that out.  It just seemed a little out of the way at the moment, since people aren't used to these options yet.  I've fixed it anyway.

clumens replied:
#1889 should fix that up.

clumens commented on tools/crm_mon_output.c:
```diff
@@ -0,0 +1,442 @@
+                       user ? " by " : "", user ? user : "",
+                       client ? " via " : "", client ? client : "",
+                       origin ? " origin " : "", origin ? origin : "",
+                       NULL);
```
I've fixed thie one up, leaving two more:
* The host_list stuff in add_extra_info.
* The g_strjoinv stuff in pe__common_output_text.

Do we have any good ways of accomplishing the same things?  I'm happy  to change them over if it's not too tedious.

kgaillot replied:
g_free() vs free() seems to be the only thing that ever gets confused, so I don't think there's a problem with glib types with specific free functions (like g_list_free(), g_strfreev()). I don't think we have a good alternative to g_strjoinv() though. The closest we have is add_list_element() which realloc()'s for each addition.

kgaillot commented on tools/crm_mon_output.c:
```diff
@@ -87,6 +87,92 @@ last_changed_string(const char *last_written, const char *user,
+
+        if (queue_str) {
+            free(queue_str);
+        }
```
Not a big deal, but free() can handle NULL, so checking is optional (it does save the function call)

kgaillot commented:
Examples should reference --output-as rather than --as-*:
crm_mon --help

Should default to stdout:
crm_mon --output-as=html

Core dumps:
crm_mon --output-as=xml --output-to=/nonexistent/file.xml
CIB_file=cts/scheduler/797.xml crm_mon -1
CIB_file=$ANY crm_mon

span id="bold" should be span class="bold"

Previously, the html title was "Cluster status" rather than the command line. Not sure which is better.

Comparing CIB_file=cts/scheduler/10-a-then-bm-b-move-a-clone.xml) crm_mon -1, previous:
    Stack: corosync
    Current DC: f20node1 (version 1.1.7-1.fc20-ce5d77c) - partition with quorum
    Last updated: Wed Sep 25 16:33:18 2019
    Last change: Sat Dec 21 11:47:45 2013 via cibadmin on f20node1
    
    2 nodes configured
    3 resources configured
    
    Node f20node1: standby (with active resources)
    Online: [ f20node2 ]
    
    Active resources:
    
     Clone Set: myclone-clone [myclone]
         Started: [ f20node1 f20node2 ]
     vm	(ocf::heartbeat:Dummy):	Started f20node1

vs new:
    Cluster Summary:
        * Stack: corosync
        * Current DC: f20node1 (version 1.1.7-1.fc20-ce5d77c) - partition with quorum
        * Last updated: Wed Sep 25 16:27:10 2019
        * Last change: Sat Dec 21 11:47:45 2013 via cibadmin origin f20node1
        * 2 nodes configured
        * 3 resources configured
    
    Node List:
        * Node f20node1: standby (with active resources)
        * Online: [ f20node2 ]
    
    Active resources:
        * Clone Set: myclone-clone [myclone]:
            * Masters: [(null) ]
            * Started: [ f20node1 f20node2 ]
        * vm	(ocf::heartbeat:Dummy):	Started f20node1

Printing the headings and using bullets definitely organizes it visually and makes it easier to grasp, although crm_mon has a constant problem with fitting in a terminal window, so we usually try to cut out everything that's not necessary. Maybe we could condense it by combining "Node list (2 nodes configured):" but I suspect that would screw up our legacy XML hacks.

We should stick with one capitalization scheme for the headings. I think four space indents is a bit much; maybe 1 or 2 for the first level, and 3 for sublevels. We definitely want to skip the Masters line if it's empty. I'm not sure whether the trailing colon on Clone Set is better.

clumens commented:
> Examples should reference --output-as rather than --as-*:
> crm_mon --help

Fixed, though this will have to go into its own patch.

> Should default to stdout:
> crm_mon --output-as=html

Fixed, though this also has to be its own patch.

> Core dumps:
> crm_mon --output-as=xml --output-to=/nonexistent/file.xml

Fixed.  This could be on a fixes branch because it's confined to lib/common/output*

> CIB_file=cts/scheduler/797.xml crm_mon -1

This is a double free caused by calling qb_log_fini from crm_exit.  Gonna take a minute to figure this one out.

> CIB_file=$ANY crm_mon

I can't reproduce this.

> span id="bold" should be span class="bold"

Fixed and merged into all the original patches.

> Previously, the html title was "Cluster status" rather than the command line. Not sure which is better.

It's better some places and worse others.  Perhaps for the HTML case alone it makes sense to override the title somewhere.

> Printing the headings and using bullets definitely organizes it visually and makes it easier to grasp, although crm_mon has a constant problem with fitting in a terminal window, so we usually try to cut out everything that's not necessary. Maybe we could condense it by combining "Node list (2 nodes configured):" but I suspect that would screw up our legacy XML hacks.

Yeah I figured this might be a problem.  I have a standalone patch that adds the blank lines between lists in crm_mon.  I could drop that patch and get several lines back.

> We should stick with one capitalization scheme for the headings.

Definitely.  I think I basically just lifted these strings from the old code.  This is a good opportunity for cleaning it up.

> I think four space indents is a bit much; maybe 1 or 2 for the first level, and 3 for sublevels. 

Easy enough.  I already have patches that change it from tabs to four spaces.  I can change that.

> We definitely want to skip the Masters line if it's empty.

Fixed.  I think this could go on a fixes branch too.

> I'm not sure whether the trailing colon on Clone Set is better.

There's a colon there because that's the start of a list.  I'd have to change it to not be a list to get rid of it.

clumens commented:
> > CIB_file=cts/scheduler/797.xml crm_mon -1
> 
> This is a double free caused by calling qb_log_fini from crm_exit. Gonna take a minute to figure this one out.

Nope, that wasn't the cause.  Fixed, though.

clumens commented:
#1893 is collecting the fixes that don't get squashed into a patch in this PR.

clumens commented:
> > Previously, the html title was "Cluster status" rather than the command line. Not sure which is better.
> 
> It's better some places and worse others. Perhaps for the HTML case alone it makes sense to override the title somewhere.

Fixed in its own patches in #1893 .

> > We should stick with one capitalization scheme for the headings.
> 
> Definitely. I think I basically just lifted these strings from the old code. This is a good opportunity for cleaning it up.

Fixed and squashed into existing patches.

> > I think four space indents is a bit much; maybe 1 or 2 for the first level, and 3 for sublevels.
> 
> Easy enough. I already have patches that change it from tabs to four spaces. I can change that.

Fixed and squashed into existing patches.

kgaillot commented:
I'm still seeing segfault for CIB_file=$ANY crm_mon

With some debug messages, I discovered it's happening because initscr() isn't being called. When the output object is created, output_format is curses, but in main()'s switch (cib->variant), cib_file forces mon_op_one_shot, and when that's set initscr() won't be called. It also resets output_format to plain at that point, but that's after out has already been created.

So we need to ensure that output_format is never changed after pcmk__output_new(). When one-shot is enabled with curses, I suppose we could either close the output object and re-open it as plain text, or allow curses to proceed anyway (which would be useful only for testing). Another question is whether CIB_file should force one-shot; the argument for it is that we likely can't or shouldn't connect to the fencer in that situation, and the argument against it is to allow curses testing of specific configurations.

BTW is there a reason the initscr() block isn't in curses_init()?

clumens commented:
> I'm still seeing segfault for CIB_file=$ANY crm_mon

Weird, I wonder why I don't see that.

> So we need to ensure that output_format is never changed after pcmk__output_new(). 

Right, and I did a lot of work on that.  I think this particular case is one I meant to come back to, but got sidetracked with something else.  The number of places we check various flags and set the output format in main is a little overwhelming.

> When one-shot is enabled with curses, I suppose we could either close the output object and re-open it as plain text, or allow curses to proceed anyway (which would be useful only for testing). Another question is whether CIB_file should force one-shot; the argument for it is that we likely can't or shouldn't connect to the fencer in that situation, and the argument against it is to allow curses testing of specific configurations.

I don't really like closing and re-opening the output object.  I think the only reason it's needed before the `crm_info("Starting %s...)` line is for printing out the version.  We don't do any setting of output_format after that point either.  So maybe there's a way to move the pcmk__output_new call to down there.  The obvious, annoying, way to do it is wrap all the preceeding code in a `if (!args->version) {` but that's a lot of extra indentation.

> BTW is there a reason the initscr() block isn't in curses_init()?

I definitely had it in there at some point but removed it.  I can't recall why.  It would be worth trying again.

clumens commented:
I still can't reproduce your segfault, so this is a bit hard to debug.  However, does the attached patch take care of things?  This should put all the output_format futzing before pcmk__output_new is called.
[0001-WIP.patch.txt](https://github.com/ClusterLabs/pacemaker/files/3682539/0001-WIP.patch.txt)

kgaillot commented:
Yep, the patch looks good and works :)

I've been testing on RHEL 7.6 -- what were you testing on? The issue may have been fixed (or worked around) in ncurses sometime after that. I plan on upgrading to 7.7 today, so I'll see if that changes anything.

A couple unrelated things I noticed:
* The  "no way to specify CGI mode" comment is no longer correct; --output-as html --output-cgi should do it.
* The daemonize check for (html || external_agent) can be more useful now as (output_file || external_agent). There's nothing wrong with updating a file in some other format, and we don't want to mention the old "--as-html" in the message anyway.

clumens commented:
> Yep, the patch looks good and works :)

I'll turn it into a real patch.

> I've been testing on RHEL 7.6 -- what were you testing on? The issue may have been fixed (or worked around) in ncurses sometime after that. I plan on upgrading to 7.7 today, so I'll see if that changes anything.

Ah, there's the difference.  I was testing on Fedora 30, but if I test on my RHEL 7.6 VM then I see the segfault.  This is fun.

>     * The  "no way to specify CGI mode" comment is no longer correct; --output-as html --output-cgi should do it.

Fixed.

>     * The daemonize check for (html || external_agent) can be more useful now as (output_file || external_agent). There's nothing wrong with updating a file in some other format, and we don't want to mention the old "--as-html" in the message anyway.

Are you referring to this:
```
            if ((output_format != mon_output_html)
                && !options.external_agent) {
                printf ("Looks like you forgot to specify one or more of: "
                        "--as-html, --external-agent\n");
                return clean_up(CRM_EX_USAGE);
            }
```

kgaillot commented:
> Are you referring to this:
> 
> ```
>             if ((output_format != mon_output_html)
>                 && !options.external_agent) {
>                 printf ("Looks like you forgot to specify one or more of: "
>                         "--as-html, --external-agent\n");
>                 return clean_up(CRM_EX_USAGE);
>             }
> ```

yep

clumens commented:
921513a0bd7e1c03360a02ec1c34af0b05c235a6 is new in the latest push, and 88ebeaa64e3f66c3e2250c9e9ca7b5160135d857 has changed quite a bit.

kgaillot commented:
Found in testing:
* The man page is empty.
* XML and HTML output is missing node_attributes, node_history, tickets, and bans sections
* With -c, tickets (e.g. cts/scheduler/tags-ticket.xml) print without a newline in text mode, and without li tags in html
* "crm_mon --as-html=$FILENAME" writes to stdout rather than $FILENAME
* Text output for cts/scheduler/anon-instance-pending.xml is missing a space after ORPHANED
* Text mode has tab indents (not sure if that's intentional)
* For "Migration Summary" (-f), nodes are printed even if they don't have any failures.
* --output-title help shows the wrong default, but that's tricky b/c crm_mon overrides the library default (maybe just drop the mention of the default from the help)
* For text mode, it might look a little cleaner if the "Last change" time had one space before it (or a space after "change") so it lines up with the "Last updated" time (not a big deal if it's nontrivial)
* With no active resources, currently it prints "Active Resources:\n\t* No active resources" whereas it might be better to print just "No active resources". But since the output will likely be small in that case, it's not important and can wait.
* Not for this PR, but thinking generally, it might be interesting to wrap long text lines automatically, especially keeping the indent for list items (possibly a new output option) (definitely a low priority, just thinking out loud)

clumens commented:
>     * The man page is empty.

Fixed in #1906.

>     * XML and HTML output is missing node_attributes, node_history, tickets, and bans sections

Fixed.

>     * With -c, tickets (e.g. cts/scheduler/tags-ticket.xml) print without a newline in text mode, and without li tags in html

Fixed.

>     * "crm_mon --as-html=$FILENAME" writes to stdout rather than $FILENAME

Fixed.

>     * Text output for cts/scheduler/anon-instance-pending.xml is missing a space after ORPHANED

Fixed.

>     * Text mode has tab indents (not sure if that's intentional)

I think this got lost in patch rebasing around the begin_list and list_item varargs stuff.  Fixed.

>     * For "Migration Summary" (-f), nodes are printed even if they don't have any failures.

Fixed.

>     * --output-title help shows the wrong default, but that's tricky b/c crm_mon overrides the library default (maybe just drop the mention of the default from the help)

Fixed by removing the mention of the default.

>     * For text mode, it might look a little cleaner if the "Last change" time had one space before it  (or a space after "change") so it lines up with the "Last updated" time (not a big deal if it's nontrivial)

It was trivial.  Fixed.

>     * With no active resources, currently it prints "Active Resources:\n\t* No active resources" whereas it might be better to print just "No active resources". But since the output will likely be small in that case, it's not important and can wait.

There's not really a straightforward way to do this.  We don't really have any functions for printing out a message across all output formats.  text_info and text_xml are the closest, but they don't function the same on other output formats.  begin_list with an immediate end_list almost works, but prints a colon at the end.  list_item outside of any list context fails due to trying to dereference a NULL pointer at its end.

>     * Not for this PR, but thinking generally, it might be interesting to wrap long text lines automatically, especially keeping the indent for list items (possibly a new output option) (definitely a low priority, just thinking out loud)

Yeah that would be really nice.  I don't know of a way to do it, but I'm sure we could find something.

kgaillot commented:
Time to break out the champagne!
